request-manager/graph/resolver.go

```diff
@@ -142,95 +144,65 @@ func (o *resolver) BuildJobChain(jobArgs map[string]interface{}) (*proto.JobChai
 		return nil, err
 	}
 
-	/* Turn it into a job chain. */
-	jc := &proto.JobChain{
-		AdjacencyList: seqGraph.Edges,
-		RequestId:     o.Request.Id,
-		State:         proto.STATE_PENDING,
-		Jobs:          map[string]proto.Job{},
-	}
-	for jobId, node := range seqGraph.Nodes {
-		bytes, err := node.Job.Serialize()
-		if err != nil {
-			return nil, err
-		}
-		job := proto.Job{
-			Type:              node.Job.Id().Type,
-			Id:                node.Job.Id().Id,
-			Name:              node.Job.Id().Name,
-			Bytes:             bytes,
-			Args:              node.Args,
-			Retry:             node.Retry,
-			RetryWait:         node.RetryWait,
-			SequenceId:        node.SequenceId,
-			SequenceRetry:     node.SequenceRetry,
-			SequenceRetryWait: node.SequenceRetryWait,
-			State:             proto.STATE_PENDING,
-		}
-		jc.Jobs[jobId] = job
-	}
-
-	return jc, nil
+	return seqGraph, nil
 }
 
-// (Recursively) build a sequence.
+// buildSequence recursively builds a sequence. If a sequence graph node represents
+// a job, buildSequence creates the corresponding job. If a sequence graph node needs
+// to be expanded, i.e. it represents anything but a job, it is recursively expanded
+// with another call to buildSequence. buildSequence also chooses the correct path
+// for conditional nodes.
 func (o *resolver) buildSequence(cfg buildSequenceConfig) (*Graph, error) {
 	sequenceName := cfg.sequenceName
 	jobArgs := cfg.jobArgs
 
-	/* Add optional and static sequence arguments to map. */
-	seq, ok := o.AllSequences[sequenceName]
-	if !ok {
-		return nil, fmt.Errorf("cannot find definition for sequence: %s", sequenceName)
-	}
-	for _, arg := range seq.Args.Optional {
-		if _, ok := jobArgs[*arg.Name]; !ok {
-			jobArgs[*arg.Name] = *arg.Default
-		}
-	}
-	for _, arg := range seq.Args.Static {
-		if _, ok := jobArgs[*arg.Name]; !ok {
-			jobArgs[*arg.Name] = *arg.Default
-		}
-	}
-
-	/* Build request graph based on sequence graph. */
-	seqGraph, ok := o.SequenceGraphs[sequenceName]
+	// Build request graph based on sequence graph.
+	seqGraph, ok := o.sequenceGraphs[sequenceName]
```

Comment:
Is this copying the seq graph? I don't think so. If that's true, then the RG is clobbering the SG?

Also, var name should be `reqGraph` (`rg` is acceptable in this case because Go likes short var names and "rg" should be unique so it's easy to search/highlight).

Reply:
Oh wait, no. Just below we do `reqGraph, err := o.newReqGraph(cfg.graphName, jobArgs)`.  Ok, so then this code should be moved closer to where it's first used. Then it makes it a little more clear to readers.

Also, code comments on this would really clarify it: we're fetching the SG only to walk its structure and build the RG in parallel. So no mutations to SG.